### DISCUSSION TITLE
[Help] Do ECDSA signatures convey signers' infomation?

### DISCUSSION THREAD

**Discussion by DisableAsync** [randombit/botan#2963] [General]
Assuming the signatures always come with their corresponding signed data. Can we find out who signed the data?

**Comment by randombit** [comment]
A signature is a binding between a message and a public key. So if you can verify the signature, you know that that particular public/private key pair was used to generate it. But that doesn't convey any real world identity, in and of itself - a public key is basically just a sequence of bits. Usually a public key <-> real world identity binding is conveyed in something like an X.509 certificate.

**Reply by DisableAsync** [reply]
My bad, by who signed the data I actually meant which key signed, not the key's owner.

IIRC in ECIES the encrypted data actually carries the encryptor's public key in its header. And it is OK because we can just use randomly generated keys to encrypt data. But for ECDSA that just won't work, there is no point signing data with random keys.

> A signature is a binding between a message and a public key. 

Does it tell us which public key it is?

**Reply by reneme** [reply]
From just the signature itself you can't extract the associated public key. However, given a list of "potential public key candidates", you can of course try each key and find the one that verifies the signature successfully.

**Reply by reneme** [reply]
Let me correct myself. @hrantzsch pointed out that there is indeed a way to recover the public key from the signature itself. See the code below for an example:

```C++
#include <iostream>
#include <memory>

#include <botan/auto_rng.h>
#include <botan/ecdsa.h>
#include <botan/hex.h>
#include <botan/ec_group.h>
#include <botan/pubkey.h>

int main(int argc, char *argv[])
{
  auto rng = std::make_unique<Botan::AutoSeeded_RNG>();
  const auto grp = Botan::EC_Group("secp256r1");

  // Create a fresh ECDSA key
  Botan::ECDSA_PrivateKey privKey(*rng, grp);

  // Sign a message with the private key
  const auto msg = Botan::hex_decode("00112233445566778899aabbccddeeff00112233445566778899aabbccddeeff");
  const auto sf = Botan::Signature_Format::IEEE_1363;
  const auto sig = Botan::PK_Signer(privKey, *rng, "Raw", sf).sign_message(msg, *rng);
  const Botan::BigInt r(sig.data(), 32);
  const Botan::BigInt s(sig.data()+32, 32);

  // Recover the public key from the signature
  // Note that one might need to 'guess' `v` which is 0 <= v < 4
  const uint8_t v = privKey.recovery_param(msg, r, s);
  Botan::ECDSA_PublicKey recovered(grp, msg, r, s, v);

  std::cout << "Public Key:           " << Botan::hex_encode(privKey.public_key_bits()) << std::endl;
  std::cout << "Recovered Public Key: " << Botan::hex_encode(recovered.public_key_bits()) << std::endl;

  return 0;
}
```

**Reply by DisableAsync** [reply]
That's one big oops!

So for anonymity of signer we should encrypt the signature as well? Guess that's gonna cost much.

**Reply by reneme** [reply]
That's correct. If you want to ensure anonymity, you should not expose plaintext signatures to a potential attacker. How to encrypt them is of course dependent on your specific use case.

**Reply by DisableAsync** [reply]
Just noticed that's "Raw EMSA" used in the example...
So if we do use an EMSA, easy public key recovery won't be possible?

**Reply by reneme** [reply]
> So if we do use an EMSA, easy public key recovery won't be possible?

Unfortunately it would still be possible. I just used the raw EMSA for simplicity. For instance `EMSA1(SHA-256)` just hashes the input data before signing the resulting hash. As there is no additional secret involved, it's easily reproducible by an attacker.

**Reply by DisableAsync** [reply]
Does changing EC group to secp521r1 help?

**Reply by reneme** [reply]
By design signatures carry a notion of "identity". Hence, different algorithms will not solve this underlying problem. Off the top of my head I'm also not aware of any "privacy preserving" signature schemes.

What's your underlying use case that requires to have a signature (i.e. strong authentication and identity) as well as anonymity?

**Reply by DisableAsync** [reply]
Thinking about some kinda TOR like decentralized networking design. Because peers don't talk to each other directly, they would need a way to verity he authenticity of the relayed (encrypted) messages they received.

And no, I'm not silly or crazy, I just want to practice something different so I can improve my skill. :P

**Reply by reneme** [reply]
So I assume you want to establish a secure channel between the endpoints that is routed through intermediate nodes. Maybe just perform a TLS handshake between the endpoints? As of TLS 1.3 the certificates (both client and server) are encrypted and wouldn't be visible to intermediate nodes.

**Reply by DisableAsync** [reply]
> Maybe just perform a TLS handshake between the endpoints? 

That sounded just like TOR (if I don't get TOR wrong)! It is a good idea. But I want something better.

If I'm ever to actually start working on the network stack, most likely the first thing I'm gonna do is implementing an instant messaging application upon it (like a PoC, but more that a PoC). And a feature I would certainly want is sending messages to or receiving from offline peers, so we wouldn't always be able to do TLS  handshakes.

**Reply by reneme** [reply]
Maybe the Signal.org blog could provide some inspiration? E.g.

* [Sealed Sender Preview](https://signal.org/blog/sealed-sender/)
* [X3DH Key Agreement Protocol](https://signal.org/docs/specifications/x3dh/)

**Reply by DisableAsync** [reply]
Sorry for the late reply, was overloaded with work. 

I'm really bad at language and might get the docs wrong, apologies if I did.

In "Sealed Sender", looks like Signal is trying to verify the senders hard. This doesn't look so good to me, and in terms of a decentralized network I don't believe it's much doable.
For "X3DH" I don't understand most parts of the docs, maybe I should just skip it and use ECIES for simplicity. (Yep, I am not a cryptography guy. Sad. )

BTW why is it so hard to find technical docs for TOR?
